numba/core/lowering.py

```diff
@@ -926,6 +938,58 @@ def _lower_call_RecursiveCall(self, fnty, expr, signature):
             )
         return res
 
+    def _lower_call_FunctionType(self, fnty, expr, signature):
+        self.debug_print("# calling first-class function type")
+        sig = types.unliteral(signature)
+        assert fnty.check_signature(signature)
+        ftype = fnty.ftype
+        argvals = self.fold_call_args(
+            fnty, sig, expr.args, expr.vararg, expr.kws,
+        )
+        func_ptr = self.__get_function_pointer(ftype, expr.func.name, sig=sig)
+        res = self.builder.call(func_ptr, argvals, cconv=fnty.cconv)
+        return res
+
+    def __get_function_pointer(self, ftype, fname, sig=None):
+        from numba.core.function_type import lower_get_wrapper_address
+
+        llty = self.context.get_value_type(ftype)
+        fstruct = self.loadvar(fname)
+        addr = self.builder.extract_value(fstruct, 0,
+                                          name='addr_of_%s' % (fname))
+
+        fptr = cgutils.alloca_once(self.builder, llty,
+                                   name="fptr_of_%s" % (fname))
+        with self.builder.if_else(
+                cgutils.is_null(self.builder, addr),
+                likely=False) as (then, orelse):
+            with then:
+                # TODO: acquire gil??
+                pyaddr = self.builder.extract_value(
+                    fstruct, 1,
+                    name='pyaddr_of_%s' % (fname))
+                # todo: check pyaddr == NULL ?
+                # cgutils.printf(self.builder, "pyaddr=%p\n", pyaddr)
+                # todo use self.api?
+                pyapi = self.context.get_python_api(self.builder)
+                addr1 = lower_get_wrapper_address(
```

Comment:
This is interesting. So it will callback into the interpreter and may trigger more compilation.  I wonder is it surviving caching or serialization (of the function). Also, the re-entrant into interpreter seems expensive and it's violating the `nopython=True`. (`print()` violates it but it's usually for debug only, while this feature is not debug only).

Reply:
Re-entering to the interpreter is triggered only when `addr==0` for some reason, most likely for some corner-case that I cannot specify atm. In fact, the condition would most likely indicate a bug.

In all unittest cases, the condition is not satisfied and the interpreter is not used.

I would be fine to remove the `lower_get_wrapper_address` call and just trigger an exception when the `addr == 0` turns to be true. What do you think?

Reply:
The `test_zero_address` test constructs a case (see `Bad2Good`) where the above code is fully used. Marking as resolved.